File: src/lib/backend/contracts.ts
/**
 * Fenerbahçe Evreni — Backend sözleşme katmanı (Faz A1)
 *
 * Tek kaynak: job tipleri, provider adları, Firestore koleksiyon adları
 * ve normalize belge şekilleri. API (A2) ve data-worker (A3+) buna uyar.
 *
 * Geriye dönük: eski koleksiyon adları (dataSyncRuns, externalPlayerMappings)
 * hâlâ rules'ta açıktır; yeni kod scrapeJobs / providerIds kullanır.
 */

// ── Koleksiyon adları ──────────────────────────────────────────────────────

export const COLLECTIONS = {
  // CMS / içerik
  players: 'players',
  matches: 'matches',
  standings: 'standings',
  articles: 'articles',
  transferReports: 'transferReports',
  polls: 'polls',
  homeSettings: 'homeSettings',
  announcements: 'announcements',
  teams: 'teams',

  // Advanced / pipeline
  advancedPlayerStats: 'advancedPlayerStats',
  advancedMatchStats: 'advancedMatchStats',

  /** Kanonik job log (yeni). Eski: dataSyncRuns */
  scrapeJobs: 'scrapeJobs',
  /** @deprecated Use scrapeJobs */
  dataSyncRuns: 'dataSyncRuns',

  /** Kanonik entity map (yeni). Eski: externalPlayerMappings */
  providerIds: 'providerIds',
  /** @deprecated Use providerIds */
  externalPlayerMappings: 'externalPlayerMappings',
} as const;

export type CollectionName = (typeof COLLECTIONS)[keyof typeof COLLECTIONS];

// ── Provider ───────────────────────────────────────────────────────────────

export const PROVIDERS = [
  'transfermarkt',
  'fbref',
  'sofascore',
  'fotmob',
  'whoscored',
  'clubelo',
  'api_football',
  'manual',
] as const;

export type Provider = (typeof PROVIDERS)[number];

// ── Job types (pipeline) ───────────────────────────────────────────────────

export const JOB_TYPES = [
  'sync_squad',
  'sync_standings',
  'sync_fixtures',
  'sync_player_season_stats',
  'sync_match_advanced',
  'sync_entity_ids',
  'health_probe',
] as const;

export type JobType = (typeof JOB_TYPES)[number];

export const JOB_STATUSES = [
  'queued',
  'running',
  'success',
  'partial',
  'failed',
  'cancelled',
] as const;

export type JobStatus = (typeof JOB_STATUSES)[number];

export const JOB_TRIGGERS = ['admin', 'scheduler', 'cli', 'api'] as const;
export type JobTrigger = (typeof JOB_TRIGGERS)[number];

/** scrapeJobs belgesi */
export interface ScrapeJobDoc {
  schemaVersion: 1;
  jobType: JobType;
  provider: Provider | Provider[];
  status: JobStatus;
  seasonKey: string;
  triggeredBy: JobTrigger;
  triggeredByEmail?: string | null;
  startedAt: string | null;
  finishedAt: string | null;
  processedCount: number;
  successCount: number;
  failedCount: number;
  recordsWritten: number;
  /** Kullanıcıya gösterilecek kısa hata; secret/stack yasak */
  errorSummary: string | null;
  /** Opsiyonel: hangi maç/oyuncu hedeflendi */
  target?: {
    matchDocumentId?: string;
    playerDocumentId?: string;
    teamSlug?: string;
  };
  meta?: Record<string, string | number | boolean | null>;
  createdAt: string;
  updatedAt: string;
}

// ── Entity map (providerIds) ───────────────────────────────────────────────

export const MAPPING_STATUSES = ['review', 'confirmed', 'rejected'] as const;
export type MappingStatus = (typeof MAPPING_STATUSES)[number];

export interface ProviderIdEntry {
  id: string;
  name?: string;
  url?: string;
}

/** providerIds belgesi — documentId = playerDocumentId (kanonik oyuncu id) */
export interface ProviderIdsDoc {
  schemaVersion: 1;
  playerDocumentId: string;
  canonicalName: string;
  slug?: string;
  aliases: string[];
  providers: Partial<Record<Provider, ProviderIdEntry>>;
  mappingStatus: MappingStatus;
  /** 0..1 — otomatik eşleşmelerde; confirmed için genelde 1 */
  confidence: number;
  verifiedBy?: string | null;
  verifiedAt?: string | null;
  createdAt: string;
  updatedAt: string;
}

// ── Advanced stats ─────────────────────────────────────────────────────────

export interface AdvancedPlayerStatsDoc {
  schemaVersion: 1;
  playerDocumentId: string;
  playerName: string;
  teamDocumentId?: string | null;
  seasonKey: string;
  competition: string;
  provider: Provider;
  providerPlayerId?: string | null;
  metrics: Record<string, number | null>;
  sourceUrl?: string | null;
  fetchedAt: string;
  createdAt: string;
  updatedAt: string;
  /** Admin kilitlediği alanlar — job ezmez */
  lockedFields?: string[];
}

export interface ShotEvent {
  x: number | null;
  y: number | null;
  xG: number | null;
  playerName?: string | null;
  playerDocumentId?: string | null;
  outcome?: string | null;
  minute?: number | null;
  teamSide?: 'home' | 'away' | null;
}

export interface AdvancedMatchStatsDoc {
  schemaVersion: 1;
  matchDocumentId: string;
  provider: Provider;
  providerMatchId?: string | null;
  homeTeam: string;
  awayTeam: string;
  competition: string;
  seasonKey: string;
  matchDate?: string | null;
  teamMetrics: {
    home: Record<string, number | null>;
    away: Record<string, number | null>;
  };
  playerMetrics?: Array<{
    playerDocumentId?: string | null;
    providerPlayerId?: string | null;
    playerName: string;
    teamSide?: 'home' | 'away' | null;
    metrics: Record<string, number | null>;
  }>;
  shotmap?: ShotEvent[];
  /** Storage path veya harici URL; ham grid ileride */
  heatmap?: {
    homeUrl?: string | null;
    awayUrl?: string | null;
    note?: string | null;
  } | null;
  sourceUrl?: string | null;
  fetchedAt: string;
  createdAt: string;
  updatedAt: string;
  lockedFields?: string[];
}

// ── Deterministic IDs ──────────────────────────────────────────────────────

export function advancedPlayerStatsId(
  playerDocumentId: string,
  seasonKey: string,
  provider: Provider
): string {
  return `${playerDocumentId}__${seasonKey}__${provider}`;
}

export function advancedMatchStatsId(
  matchDocumentId: string,
  provider: Provider
): string {
  return `${matchDocumentId}__${provider}`;
}

// ── API envelope (A2 ile kullanılacak) ─────────────────────────────────────

export interface ApiSuccess<T> {
  success: true;
  data: T;
  meta?: {
    fetchedAt?: string;
    provider?: Provider | string;
    count?: number;
  };
}

export interface ApiError {
  success: false;
  message: string;
  code?: string;
}

export type ApiResponse<T> = ApiSuccess<T> | ApiError;

export function isJobType(v: string): v is JobType {
  return (JOB_TYPES as readonly string[]).includes(v);
}

export function isProvider(v: string): v is Provider {
  return (PROVIDERS as readonly string[]).includes(v);
}

File: data-worker/contracts.py
"""
Fenerbahçe Evreni — data-worker sözleşme katmanı (Faz A1)

TypeScript: src/lib/backend/contracts.ts ile senkron tutulmalı.
Job runner (A3+) bu sabitleri ve belge şekillerini kullanır.
"""

from __future__ import annotations

from typing import Any, Literal

# ── Koleksiyon adları ───────────────────────────────────────────────────────

COLLECTIONS = {
    "players": "players",
    "matches": "matches",
    "standings": "standings",
    "advanced_player_stats": "advancedPlayerStats",
    "advanced_match_stats": "advancedMatchStats",
    "scrape_jobs": "scrapeJobs",
    "provider_ids": "providerIds",
    # Geriye dönük
    "data_sync_runs": "dataSyncRuns",
    "external_player_mappings": "externalPlayerMappings",
}

PROVIDERS = (
    "transfermarkt",
    "fbref",
    "sofascore",
    "fotmob",
    "whoscored",
    "clubelo",
    "api_football",
    "manual",
)

JOB_TYPES = (
    "sync_squad",
    "sync_standings",
    "sync_fixtures",
    "sync_player_season_stats",
    "sync_match_advanced",
    "sync_entity_ids",
    "health_probe",
)

JOB_STATUSES = (
    "queued",
    "running",
    "success",
    "partial",
    "failed",
    "cancelled",
)

JOB_TRIGGERS = ("admin", "scheduler", "cli", "api")

MAPPING_STATUSES = ("review", "confirmed", "rejected")

SCHEMA_VERSION = 1


def advanced_player_stats_id(player_document_id: str, season_key: str, provider: str) -> str:
    return f"{player_document_id}__{season_key}__{provider}"


def advanced_match_stats_id(match_document_id: str, provider: str) -> str:
    return f"{match_document_id}__{provider}"


def new_scrape_job(
    *,
    job_type: str,
    provider: str | list[str],
    season_key: str,
    triggered_by: str = "cli",
    triggered_by_email: str | None = None,
    target: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """scrapeJobs belgesi iskeleti (ISO timestamps runner doldurur)."""
    if job_type not in JOB_TYPES:
        raise ValueError(f"Geçersiz jobType: {job_type}")
    providers = provider if isinstance(provider, list) else [provider]
    for p in providers:
        if p not in PROVIDERS:
            raise ValueError(f"Geçersiz provider: {p}")
    if triggered_by not in JOB_TRIGGERS:
        raise ValueError(f"Geçersiz triggeredBy: {triggered_by}")

    return {
        "schemaVersion": SCHEMA_VERSION,
        "jobType": job_type,
        "provider": provider,
        "status": "queued",
        "seasonKey": season_key,
        "triggeredBy": triggered_by,
        "triggeredByEmail": triggered_by_email,
        "startedAt": None,
        "finishedAt": None,
        "processedCount": 0,
        "successCount": 0,
        "failedCount": 0,
        "recordsWritten": 0,
        "errorSummary": None,
        "target": target or {},
        "meta": {},
        "createdAt": None,  # runner ISO UTC yazar
        "updatedAt": None,
    }


def is_job_type(value: str) -> bool:
    return value in JOB_TYPES


def is_provider(value: str) -> bool:
    return value in PROVIDERS
